src/Capsule_TEST.cc

```diff
@@ -0,0 +1,127 @@
+ *
+*/
+#include <gtest/gtest.h>
+#include <cmath>
+#include <iostream>
+
+#include "ignition/math/Capsule.hh"
+
+using namespace ignition;
+
+/////////////////////////////////////////////////
+TEST(CapsuleTest, Constructor)
+{
+  // Default constructor
+  {
+    math::Capsuled capsule;
+    EXPECT_DOUBLE_EQ(0.0, capsule.Length());
+    EXPECT_DOUBLE_EQ(0.0, capsule.Radius());
+    EXPECT_EQ(math::Material(), capsule.Mat());
+
+    math::Capsuled capsule2;
+    EXPECT_EQ(capsule, capsule2);
+  }
+
+  // Length and radius constructor
+  {
+    math::Capsuled capsule(1.0, 2.0);
+    EXPECT_DOUBLE_EQ(1.0, capsule.Length());
+    EXPECT_DOUBLE_EQ(2.0, capsule.Radius());
+    EXPECT_EQ(math::Material(), capsule.Mat());
+
+    math::Capsuled capsule2(1.0, 2.0);
+    EXPECT_EQ(capsule, capsule2);
+  }
+
+  // Length, radius, mat
+  {
+    math::Capsuled capsule(1.0, 2.0,
+        math::Material(math::MaterialType::WOOD));
+    EXPECT_DOUBLE_EQ(1.0, capsule.Length());
+    EXPECT_DOUBLE_EQ(2.0, capsule.Radius());
+    EXPECT_EQ(math::Material(math::MaterialType::WOOD), capsule.Mat());
+
+    math::Capsuled capsule2(1.0, 2.0,
+        math::Material(math::MaterialType::WOOD));
+    EXPECT_EQ(capsule, capsule2);
+  }
+}
+
+//////////////////////////////////////////////////
+TEST(CapsuleTest, Mutators)
+{
+  math::Capsuled capsule;
+  EXPECT_DOUBLE_EQ(0.0, capsule.Length());
+  EXPECT_DOUBLE_EQ(0.0, capsule.Radius());
+  EXPECT_EQ(math::Material(), capsule.Mat());
+
+  capsule.SetLength(100.1);
+  capsule.SetRadius(.123);
+  capsule.SetMat(math::Material(math::MaterialType::PINE));
+
+  EXPECT_DOUBLE_EQ(100.1, capsule.Length());
+  EXPECT_DOUBLE_EQ(.123, capsule.Radius());
+  EXPECT_EQ(math::Material(math::MaterialType::PINE), capsule.Mat());
+}
+
+//////////////////////////////////////////////////
+TEST(CapsuleTest, VolumeAndDensity)
+{
+  double mass = 1.0;
+  math::Capsuled capsule(1.0, 0.001);
+  double expectedVolume = (IGN_PI * std::pow(0.001, 2) * (1.0 + 4./3. * 0.001));
+  EXPECT_DOUBLE_EQ(expectedVolume, capsule.Volume());
+
+  double expectedDensity = mass / expectedVolume;
+  EXPECT_DOUBLE_EQ(expectedDensity, capsule.DensityFromMass(mass));
+
+  // Bad density
+  math::Capsuled capsule2;
+  EXPECT_GT(0.0, capsule2.DensityFromMass(mass));
+}
+
+//////////////////////////////////////////////////
+TEST(CapsuleTest, Mass)
+{
+  double mass = 2.0;
+  double l = 2.0;
+  double r = 0.1;
+  math::Capsuled capsule(l, r);
+  capsule.SetDensityFromMass(mass);
+
+  const double cylinderVolume = IGN_PI * r*r * l;
+  const double sphereVolume = IGN_PI * 4. / 3. * r*r*r;
+  const double volume = cylinderVolume + sphereVolume;
+  const double cylinderMass = mass * cylinderVolume / volume;
+  const double sphereMass = mass * sphereVolume / volume;
+
+  // expected values based on formula used in Open Dynamics Engine
+  // https://bitbucket.org/odedevs/ode/src/0.16.2/ode/src/mass.cpp#lines-148:153
+  double ixxIyy = (1/12.0) * cylinderMass * (3*r*r + l*l)
```

Comment:
I checked this against DART (https://github.com/dartsim/dart/blob/master/dart/dynamics/CapsuleShape.cpp#L123) and there is a discrepancy, but I think what you have here is correct. The referenced page (https://www.gamedev.net/tutorials/_/technical/math-and-physics/capsule-inertia-tensor-r3856/) aligns with the formula here except the `0.25*l*l` term here vs `0.5*l*l` term there. Scrolling down further, I see a comment that it should actually be `0.25*l*l`.

Reply:
it shows `H^2 m / 4` in equation 12 on that page; I think there was just a transcription error in equation 14

Reply:
I picked ODE because I think it's right, though it's not authoritative and doesn't have any justification. I could add a link to that gamedev.net page as well if you think that's a good reference, despite its inconsistency

Reply:
The ODE reference is fine. I just checked DART to verify and saw it was different.